### PR TITLE
LTE link controller extensions

### PR DESCRIPTION
**Pull request by jtguggedal**
### drivers: lte_lc: Add API to get network registration status

This patch adds an API to get the current network registration
status.
The LTE link controller is changed to use the new parsing
function for registration status internally as well.

---

### drivers: lte_lc: Add API for setting and getting system mode

Adds APIs for getting and setting system mode.
Also adds utility function for checking if a returned AT
response is the same as expected.

---

### drivers: lte_lc: Add API for getting functional mode

Adds API for getting the modem's current functional mode.

### CODE REVIEW

File: drivers/lte_link_control/lte_lc.c [hunk 1]
```diff
@@ -19,14 +19,37 @@
...
+#define AT_CEREG_READ				"AT+CEREG?"
+#define AT_CEREG_RESPONSE_PREFIX		"+CEREG"
+#define AT_CEREG_PARAMS_COUNT			10
+#define AT_CEREG_REG_STATUS_INDEX		1
+#define AT_CEREG_ACTIVE_TIME_INDEX		8
+#define AT_CEREG_TAU_INDEX			9
+#define AT_CEREG_RESPONSE_MAX_LEN		80
+#define AT_XSYSTEMMODE_READ			"AT%XSYSTEMMODE?"
+#define AT_XSYSTEMMODE_RESPONSE_PREFIX		"%XSYSTEMMODE"
+#define AT_XSYSTEMMODE_PROTO			"AT%%XSYSTEMMODE=%d,%d,%d,%d"
```
Comment (Rallare) [inline drivers/lte_link_control/lte_lc.c]: Just a generic note on such proto defines, these should be used with care as %d assumes that the value going in is only one byte ASCII (ie: 0 to 9). If the input (or expected return) is two bytes, its going to be an overflow if you use buf[sizeof(AT_XSYSTEMMODE_PROTO)];

Comment (jtguggedal) [reply drivers/lte_link_control/lte_lc.c]: I see your point. As you mention, the buffer size is sufficient here, but happy to take suggestions on how to improve this.

File: drivers/lte_link_control/lte_lc.c [hunk 2]
```diff
@@ -423,6 +421,116 @@ int lte_lc_edrx_req(bool enable)
...
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+static int parse_nw_reg_status(char *at_response,
+			       enum lte_lc_nw_reg_status *status)
+{
+	int err, reg_status;
+	struct at_param_list resp_list = {0};
+	char  response_prefix[sizeof(AT_CEREG_RESPONSE_PREFIX)] = {0};
+	size_t response_prefix_len = ARRAY_SIZE(response_prefix);
```
Comment (rlubos) [inline drivers/lte_link_control/lte_lc.c]: Won't `sizeof` be more natural for this case than `ARRAY_SIZE`?

File: drivers/lte_link_control/lte_lc.c [hunk 3]
```diff
@@ -531,6 +560,151 @@ int lte_lc_nw_reg_status_get(enum lte_lc_nw_reg_status *status)
...
+	return err;
+}
+
+int lte_lc_system_mode_get(enum lte_lc_system_mode *mode)
+{
+	int err, bitmask = 0;
+	struct at_param_list resp_list = {0};
+	char response[AT_XSYSTEMMODE_RESPONSE_MAX_LEN] = {0};
+	char response_prefix[sizeof(AT_XSYSTEMMODE_RESPONSE_PREFIX)] = {0};
+	size_t response_prefix_len = ARRAY_SIZE(response_prefix);
```
Comment (rlubos) [inline drivers/lte_link_control/lte_lc.c]: `sizeof` instead of `ARRAY_SIZE`?

File: drivers/lte_link_control/lte_lc.c [hunk 4]
```diff
@@ -423,6 +436,338 @@ int lte_lc_edrx_req(bool enable)
...
+	return err;
+}
+
+int lte_lc_func_mode_get(enum lte_lc_func_mode *mode)
+{
+	int err, resp_mode;
+	struct at_param_list resp_list = {0};
+	char response[AT_CFUN_RESPONSE_MAX_LEN] = {0};
+	char response_prefix[sizeof(AT_CFUN_RESPONSE_PREFIX)] = {0};
+	size_t response_prefix_len = ARRAY_SIZE(response_prefix);
```
Comment (rlubos) [inline drivers/lte_link_control/lte_lc.c]: `sizeof`?

File: drivers/lte_link_control/lte_lc.c (hunk 2 above)
Comment (jtguggedal) [reply drivers/lte_link_control/lte_lc.c]: Yep, fixed all of them. Thanks

File: drivers/lte_link_control/lte_lc.c [hunk 5]
```diff
@@ -423,6 +436,338 @@ int lte_lc_edrx_req(bool enable)
...
+ * @param status Pointer to where the registration status is stored.
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+static int parse_nw_reg_status(const char *at_response,
+			       enum lte_lc_nw_reg_status *status)
+{
+	int err, reg_status;
+	struct at_param_list resp_list = {0};
+	char  response_prefix[sizeof(AT_CEREG_RESPONSE_PREFIX)] = {0};
```
[the remark this answers is not in the document]
Comment (jtguggedal) [reply drivers/lte_link_control/lte_lc.c]: Yeah, this is lazy null-terminating by me. I'll keep it in mind, thanks.

**Comment by rlubos** [comment]
@Rallare Any other comments? If not, go ahead and approve :)

File: drivers/lte_link_control/lte_lc.c [hunk 6]
```diff
@@ -423,6 +436,338 @@ int lte_lc_edrx_req(bool enable)
...
+/**@brief Parses an AT command response, and returns the current network
+ *	  registration status if it's available in the string.
+ *
+ * @param at_response Pointer to buffer with AT response.
+ * @param status Pointer to where the registration status is stored.
+ *
+ * @return Zero on success or (negative) error code otherwise.
+ */
+static int parse_nw_reg_status(const char *at_response,
+			       enum lte_lc_nw_reg_status *status)
```
Comment (Rallare) [inline drivers/lte_link_control/lte_lc.c]: "status" needs to be cheked against NULL - "at_response" is checked by at_parser_max_params_from_str()

File: drivers/lte_link_control/lte_lc.c [hunk 7]
```diff
@@ -423,6 +436,338 @@ int lte_lc_edrx_req(bool enable)
...
+	err = at_cmd_write(cmd, NULL, 0, NULL);
+	if (err) {
+		LOG_ERR("Could not send AT command, err: %d", err);
+	}
+
+exit:
+	return err;
+}
+
+int lte_lc_system_mode_get(enum lte_lc_system_mode *mode)
```
Comment (Rallare) [inline drivers/lte_link_control/lte_lc.c]: if mode == NULL

File: drivers/lte_link_control/lte_lc.c [hunk 8]
```diff
@@ -423,6 +436,338 @@ int lte_lc_edrx_req(bool enable)
...
+		break;
+	}
+
+clean_exit:
+	at_params_list_free(&resp_list);
+
+	return err;
+}
+
+int lte_lc_func_mode_get(enum lte_lc_func_mode *mode)
```
Comment (Rallare) [inline drivers/lte_link_control/lte_lc.c]: check if mode == NULL

File: drivers/lte_link_control/lte_lc.c (hunk 6 above)
Comment (jtguggedal) [reply drivers/lte_link_control/lte_lc.c]: Fixed (added check for both, though a tiny bit redundant)

File: drivers/lte_link_control/lte_lc.c (hunk 7 above)
Comment (jtguggedal) [reply drivers/lte_link_control/lte_lc.c]: Fixed

**Comment by jtguggedal** [comment]
Added `NULL` checks per @Rallare's review. Relevant diff from force push here: https://github.com/NordicPlayground/fw-nrfconnect-nrf/compare/64814e3d7c1b42bd19bd378c004d6bd58ac87873..6b9acb495772e8440188004dc024d76a34710adc#diff-9663629f3b556ca2e6c43056b89e09f3

**Pull request merged**